larq_compute_engine/core/indirect_bgemm/kernel.h

```diff
@@ -1,123 +1,195 @@
-    // `if constexpr` (C++17), this causes a compile error despite the fact that
-    // we know that within this if block DstScalar must be float or int8.
-    using DS =
-        typename std::conditional<is_float_or_int8, DstScalar, float>::type;
-    using KFn = typename IndirectBGEMMKernel<DstScalar>::MicroKernelFunction;
-
-    if (bitpacked_input_shape.Dims(3) % 4 == 0) {
-      if (bitpacked_input_shape.Dims(3) > 4) {
-        return {(KFn*)&kernel_8x4x4_aarch64::RunKernel<DS, true>, 8, 4, 4};
-      } else {
-        return {(KFn*)&kernel_8x4x4_aarch64::RunKernel<DS, false>, 8, 4, 4};
-      }
-    } else if (bitpacked_input_shape.Dims(3) % 2 == 0) {
-      if (bitpacked_input_shape.Dims(3) > 2) {
-        return {(KFn*)&kernel_8x4x2_aarch64::RunKernel<DS, true>, 8, 4, 2};
-      } else {
-        return {(KFn*)&kernel_8x4x2_aarch64::RunKernel<DS, false>, 8, 4, 2};
-      }
-    } else {
-      if (bitpacked_input_shape.Dims(3) > 1) {
-        return {(KFn*)&kernel_8x4x1_aarch64::RunKernel<DS, true>, 8, 4, 1};
-      } else {
-        return {(KFn*)&kernel_8x4x1_aarch64::RunKernel<DS, false>, 8, 4, 1};
+  const std::int32_t input_depth;
+  const std::int32_t output_channels;
+  const std::int32_t filter_size;
+  const std::int32_t groups;
+  const std::int32_t num_output_pixels;
+
+  std::vector<TBitpacked> packed_weights;
+  std::vector<const TBitpacked*> indirection_buffer;
+  std::vector<TBitpacked> zero_buffer;
+
+  Kernel(const std::int32_t block_size_output_channels,
+         const std::int32_t block_size_pixels,
+         const std::int32_t block_size_depth,
+         const bconv2d::BConv2DParams* bconv2d_params,
+         const RuntimeShape& bitpacked_input_shape,
+         const RuntimeShape& output_shape)
+      : block_size_output_channels(block_size_output_channels),
+        block_size_pixels(block_size_pixels),
+        block_size_depth(block_size_depth),
+        input_depth(bitpacked_input_shape.Dims(3)),
+        output_channels(bconv2d_params->channels_out),
+        filter_size(bconv2d_params->filter_height *
+                    bconv2d_params->filter_width),
+        groups(bconv2d_params->groups),
+        num_output_pixels(bitpacked_input_shape.Dims(0) * output_shape.Dims(1) *
+                          output_shape.Dims(2)) {}
+
+  /**
+   * Pack the weights in the correct order. This procedure is (heavily)
+   * adapted from the following XNNPack function:
+   * https://github.com/google/XNNPACK/blob/80a8ac59849bfdae8d2e1409f5642baa502c0b9e/src/packing.c#L429-L484
+   */
+  void PackWeights(const TBitpacked* weights_ptr) {
+    const std::int32_t input_depth_per_group = input_depth / groups;
+    const std::int32_t output_channels_per_group = output_channels / groups;
+    const std::int32_t rounded_up_output_channels_per_group =
+        block_size_output_channels *
+        ((output_channels_per_group + block_size_output_channels - 1) /
+         block_size_output_channels);
```

Comment:
The functions `PackWeights` and `FillIndirectionBuffer` are now member functions, whereas they were previously they were defined in `prepare.h`.